projects/06/hack_assembler/hack_assembler/parser.py

```diff
@@ -0,0 +1,72 @@
+from .exceptions import LineParseException
+from .instruction import parse_instruction
+from .consts import LABEL_REGEX
+
+
+class SymbolTable(object):
+
+    def __init__(self):
+        self._symbol_table = {
+            'SP': 0,
+            'LCL': 1,
+            'ARG': 2,
+            'THIS': 3,
+            'THAT': 4,
+            'SCREEN': 16384,
+            'KBD': 24576
+        }
+        self._symbol_table.update({f'R{i}': i for i in range(16)})
+        self._current_variable_location = 16
+
+    def add_label(self, label, location):
+        self._symbol_table[label] = location
+
+    def add_variable(self, var):
+        self._symbol_table[var] = self._current_variable_location
+        self._current_variable_location += 1
+
+    def __contains__(self, item):
+        return item in self._symbol_table
+
+    def __getitem__(self, item):
+        return self._symbol_table[item]
+
+    @staticmethod
+    def build_symbol_table(code):
+        symbol_table = SymbolTable()

```

Comment:
its seems odd that going over the code is implemented twice
this is the first time, and 2nd time is inside ```parse_data``` itself.

the first time uses regex and does not handle language errors
the 2nd time uses the parser and does correctly handle exceptions in the language

its also wierd, abstraction wise, that build_symbol_table returns the "clean" code, since building the symbol table requires two passes and cannot be fully complete here. 
I suggest we separate the abstractions, the symbol table is initialized with the built ins, and then constructed by the parser in two phases. 

phase 1: go over the code, for each label instruction, call ```add_label```
phase 2: go over the code, ignoring label instructions, if encountered a symbol in A instruction, allocate in symbol table (like you're already doing now)

This will also allow you to reuse the parser loop (by adding the label syntax to the parser spec) and error handling

Reply:
I don't see a problem with iteration being implemented in 2 different places since the purpose of the iteration is very different and the cost to implement it in a generic way is unreadable code.

I agree with the abstraction problem so I will move the code cleaning to a better place.

Reply:
"implement it in a generic way is unreadable code." 
if you implement it badly then sure